src/RadixDlt.NetworkGateway.PostgresIntegration/LedgerExtension/ValidatorProcessor.cs

```diff
@@ -0,0 +1,236 @@
+ * Work. You assume all risks associated with Your use of the Work and the exercise of
+ * permissions under this License.
+ */
+
+using NpgsqlTypes;
+using RadixDlt.NetworkGateway.Abstractions;
+using RadixDlt.NetworkGateway.Abstractions.Model;
+using RadixDlt.NetworkGateway.Abstractions.Numerics;
+using RadixDlt.NetworkGateway.PostgresIntegration.Models;
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+using CoreModel = RadixDlt.CoreApiSdk.Model;
+
+namespace RadixDlt.NetworkGateway.PostgresIntegration.LedgerExtension;
+
+internal record struct ValidatorPublicKeyLookup(long ValidatorEntityId, PublicKeyType PublicKeyType, ValueBytes PublicKey);
+
+internal record ValidatorActiveSet(long Epoch, IDictionary<ValidatorPublicKeyLookup, TokenAmount> ValidatorSet, long StateVersion);
+
+internal class ValidatorProcessor
+{
+    private readonly ProcessorContext _context;
+    private readonly ReferencedEntityDictionary _referencedEntities;
+
+    private List<ValidatorActiveSet> _changes = new();
+
+    private Dictionary<ValidatorPublicKeyLookup, long> _seenPublicKeys = new();
+    private Dictionary<ValidatorPublicKeyLookup, ValidatorPublicKeyHistory> _existingPublicKeys = new();
+
+    private List<ValidatorPublicKeyHistory> _publicKeysToAdd = new();
+    private List<ValidatorActiveSetHistory> _activeSetsToAdd = new();
+
+    public ValidatorProcessor(ProcessorContext context, ReferencedEntityDictionary referencedEntities)
+    {
+        _context = context;
+        _referencedEntities = referencedEntities;
+    }
+
+    public void VisitUpsert(CoreModel.Substate substateData, ReferencedEntity referencedEntity, long stateVersion, long? passingEpoch)
+    {
+        if (substateData is CoreModel.ConsensusManagerRegisteredValidatorsByStakeIndexEntrySubstate entry)
+        {
+            var av = entry.Value.ActiveValidator;
+            var lookup = new ValidatorPublicKeyLookup(_referencedEntities.Get((EntityAddress)av.Address).DatabaseId, av.Key.KeyType.ToModel(), av.Key.GetKeyBytes());
+
+            _seenPublicKeys.TryAdd(lookup, stateVersion);
+        }
+
+        if (substateData is CoreModel.ValidatorFieldStateSubstate state)
+        {
+            var lookup = new ValidatorPublicKeyLookup(referencedEntity.DatabaseId, state.Value.PublicKey.KeyType.ToModel(), state.Value.PublicKey.GetKeyBytes());
+
+            _seenPublicKeys.TryAdd(lookup, stateVersion);
+        }
+
+        if (substateData is CoreModel.ConsensusManagerFieldCurrentValidatorSetSubstate validatorSet)
+        {
+            if (!passingEpoch.HasValue)
+            {
+                throw new InvalidOperationException("ConsensusManagerFieldCurrentValidatorSetSubstate can't be processed unless epoch change gets detected");
+            }
+
+            var activeSet = new Dictionary<ValidatorPublicKeyLookup, TokenAmount>();
```

Comment:
that name is missleading, can we rename it to i.e validatorStakeDictionary? or sthn else that would indicate we hold stake amount per validator here?

Reply:
Clarified.